apps/system/js/audio_channel_manager.js

```diff
@@ -0,0 +1,370 @@
+
+    /**
+     * Initial the module.
+     */
+    _start: function() {
+      this._activeAudioChannels = new Map();
+      this._interruptedAudioChannels = new Map();
+      this._audioChannelsResumeWhenOtherEnds = [];
+      this.readSetting('vibration.enabled').then((value) => {
+        this._isVibrateEnabled = value;
+      });
+      this.debug('Start Audio Channel Manager');
+    },
+
+    /**
+     * Handle the audio chanel when it is active or in inactive.
+     *
+     * @param {Event} evt The event to handle.
+     */
+    _handle_audiochannelstatechanged: function(evt) {
+      var audioChannel = evt.detail.audioChannel;
+      this.debug('Audio channel state is ' + audioChannel.states.active);
+      this._manageAudioChannels(audioChannel);
+    },
+
+    /**
+     * Handle the audio chanel when the app is in foreground or background.
+     */
+    _handle_hierarchytopmostwindowchanged: function() {
+      if (this._topMostWindow) {
+        // Normal channel could not play in background.
+        this.debug(this._topMostWindow.name + ' is closed');
+        var audioChannel = this._topMostWindow.getAudioChannel('normal');
+        if (audioChannel && audioChannel.states.playing) {
+          this._handleAudioChannel({
+            isAllowedToPlay: false,
+            audioChannel: audioChannel
+          })
+          .then(this._resumeBackgroundAudioChannel.bind(this));
+        }
+      }
+      this._topMostWindow = Service.query('getTopMostWindow');
+      // Resume the foreground app's audio channels.
+      this.debug(this._topMostWindow.name + ' is opened');
+      this._resumeForegroundAudioChannels(this._topMostWindow);
+    },
+
+    /**
+     * Register the window's audio channels to be managed
+     * by AudioChannelManager when the window is created.
+     *
+     * @param {Event} evt The event to handle.
+     */
+    _handle_windowcreated: function(evt) {
+      var app = evt.detail;
+      app.registerAudioChannels();
+    },
+
+    /**
+     * Remove the window's audio channels from
+     * `_activeAudioChannels`, `_interruptedAudioChannels`,
+     * and `_audioChannelsResumeWhenOtherEnds`
+     * when the window is terminated.
+     *
+     * @param {Event} evt The event to handle.
+     */
+    _handle_windowterminated: function(evt) {
+      if (this._activeAudioChannels.size ||
+          this._interruptedAudioChannels.size) {
+        var app = evt.detail;
+        app.browser.element.allowedAudioChannels.forEach((audioChannel) => {
+          var id = app.getAudioChannel(audioChannel.name).instanceID;
+          this._activeAudioChannels.delete(id);
+          var element = this._interruptedAudioChannels.get(id);
+          if (element) {
+            this._interruptedAudioChannels.delete(id);
+            var index = element.backgroundResumeIndex;
+            index != null &&
+              this._audioChannelsResumeWhenOtherEnds.splice(index, 1);
+          }
+        });
+      }
+    },
```

Comment:
Again, we don't need to know "any window is terminated". We could have AudioChannelController to listen to _terminated event from its app and fire audiochannelterminated event to notify AudioChannelManager.